ggml/src/ggml-cuda/mmvq.cu

```diff
@@ -190,12 +190,28 @@ static __global__ void mul_mat_vec_q(
 
     const uint32_t channel_bias = ids ? channel_x : channel_dst;
 
+    float x_biases[ncols_dst][rows_per_cuda_block]    = { { 0.0f } };
+    float gate_biases[ncols_dst][rows_per_cuda_block] = { { 0.0f } };
```

Comment:
```suggestion
    float x_biases[ncols_dst]    = { { 0.0f } };
    float gate_biases[ncols_dst] = { { 0.0f } };
```

These arrays are larger than they need to be. During the matrix multiplication each thread accumulates a partial sum in `tmp`. At the end of the kernel those partial sums are reduced and each thread holds the same value (ignoring differences in floating point rounding error). Then, because each thread effectively holds the same data the writeback across `rows_per_cuda_block` is parallelized. But each thread still only needs a single value from `x_bias` and `gate_bias`.

I don't think the compiler can optimize out the unused values since the indexing via `threadIdx.x` is not known at compile time.